### ISSUE TITLE
mcMMO incompatibility: Invincible Mobs

### ISSUE THREAD

**Issue by WhatsTheBadNews** [GC-spigot/AdvancedEnchantments#2012]
## Details

**Describe the bug**
Mobs can be made invincible, but still be able to attack the player. Sometimes they look normal and can't take damage. Other times they are stuck in a dying animation.

This issue happens pretty rarely for each individual on my server because the factors involved only come together by rare chance. But overall it is a pretty common occurrence on my server due to the amount of players and the amount of mobs being grinded. With some tweaks tome some settings for testing, it can be recreated reliably.

When you use swords with mcMMO, the Swords skill grants a chance of the "Rupture" effect which does periodic-damage that is non-lethal occurring twice a second and bypasses armor.

I don't know the exact mechanics going on behind this, but I believe some combination of 1.) the mob having low enough health 2.) the mob currently "Rupturing" (which is supposed to be non-lethal, so perhaps it prevents the death in a bugged way?) 3.) getting effected by "EXTRA_DAMAGE" effect at the same time causes the issue. I can very reliably recreate this issue after a lot of testing.

TLDR; Rupture and EXTRA_DAMAGE are the important factors.

I believe it is related to this issue: https://github.com/mcMMO-Dev/mcMMO/issues/4507

NOTE: I've reported this issue to mcMMO as well as it may be something that needs to be fixed on their end.
https://github.com/mcMMO-Dev/mcMMO/issues/4537
But I hope you guys can take a look at it and let me know just to be sure. Thanks 😃 

**To Reproduce**
1.) Install mcMMO 2.1.196. You don't need to buy it. You can compile it for free here https://github.com/mcMMO-Dev/mcMMO
2.) Go to the `advanced.yml` for mcMMO and make sure the settings are set as such (this is to make testing easier, it can be done accomplished with the default settings, it is just more chance based)
```
    Swords:
        Rupture:
            Rupture_Mechanics:
                # This is % chance, 15 would mean 15% percent of the time
                Chance_To_Apply_On_Hit:
                    Rank_1: 100
                    Rank_2: 100
                    Rank_3: 100
                    Rank_4: 100
```
3.) Add this enchantment to your AdvancedEnchantments `enchantments.yml` (the settings can be a bit different and it still work, but this makes the testing easier)
```
criticalhit:
  display: '%group-color%Critical Hit'
  description: "A chance to land\na Critical Hit."
  applies-to: 'Weapons'
  type: 'ATTACK;ATTACK_MOB'
  group: 'LEGENDARY'
  applies:
  - ALL_SWORD
  - ALL_AXE
  levels:
    1:
      chance: 100
      cooldown: 1
      effects:
      - 'EXTRA_DAMAGE:1.3 %victim%'
      - 'MESSAGE:&6&l** CRITICAL HIT ** %attacker%'
```
4.) Start your server and run the command `/mmoedit [player] Swords 1000`
5.) Get a diamond sword with sharpness 5
6.) Get a wooden sword and enchant it with Critical Hit 1

-- FROM HERE, YOU CAN JUST WATCH THE VIDEOS, BUT I WILL WRITE OUT THE STEPS AS WELL FOR CLARITY --

7.) Spawn a ravager. (This glitch happens with all different types of mobs, however the ravager is the one I have optimized testing to consistently recreate the issue)
8.) Attack with the diamond sword until the mcMMO health bar shows 2 hearts
![image](https://user-images.githubusercontent.com/29821149/119913321-42c68880-bf23-11eb-8303-8c2eb95f2625.png)
9.) Immediately switch to the wooden sword with Critical Hit and attack
10.) Let the Rupture play out and you will now have an invincible mob

**Screenshots / Video**
EDITED VERSION WITH EXPLANATIONS: https://www.youtube.com/watch?v=Cohl5sMvgPQ
UNEDITED VERSION, EASIER TO SEE EVERYTHING: https://www.youtube.com/watch?v=kKx6r1eH11E

## Server Information
- "/ae plinfo" link: https://paste.md-5.net/ehagurenup
- Server log: https://gist.github.com/WhatsTheBadNews/71d7da37b2db4cf71e9ff7ed355145ae

**Comment by Potrock** [comment]
This happens with MythicMobs mobs too, but MythicMobs **does** fire a MythicMobDeathEvent. The mob will just be invincible afterwards.

**Comment by Insprill** [comment]
In 8.5.7 I redid how we apply damage, and added a `use-experimental-damage` option to change it in a different way. Can y'all test and see if either of those makes a difference to this?

**Comment by hkkongou** [comment]
> In 8.5.7 I redid how we apply damage, and added a `use-experimental-damage` option to change it in a different way. Can y'all test and see if either of those makes a difference to this?

i have tested for 2 days in 3 servers with experimental-damage on
my players reported that after this option is on, the number of immortal mobs have been greatly reduced, however, there are still some immortal mobs, but very less.

i have mythicmobs and mcmmo installed

**Comment by MattTek** [comment]
I have the same problem with mcMMO's rupture and AdvancedEnchants on my server. Using `use-experimental-damage` made the issue pretty much go away, but after 10-12hrs we ran into immortal mobs again, similar to @hkkongou.

**Comment by Insprill** [comment]
In the next update i made some more changes to how use-experimental-damage is handled. Once it comes out please give it a go and see if it fixes it.

**Comment by Insprill** [comment]
> In the next update i made some more changes to how use-experimental-damage is handled. Once it comes out please give it a go and see if it fixes it.
@hkkongou @MattTek

**Comment by hkkongou** [comment]
> > In the next update i made some more changes to how use-experimental-damage is handled. Once it comes out please give it a go and see if it fixes it.
> 
> @hkkongou @MattTek

sure 
let me try try for a few days

**Comment by NCSV** [comment]
> > In the next update i made some more changes to how use-experimental-damage is handled. Once it comes out please give it a go and see if it fixes it.
> 
> @hkkongou @MattTek

[2021-09-06-4.log](https://github.com/GC-spigot/AdvancedEnchantments/files/7116565/2021-09-06-4.log)
![1](https://user-images.githubusercontent.com/25452835/132231947-d0f7eb17-030d-4408-be60-eb51452bb73a.PNG)

spigot 1.16.5
mythicmobs / advancedenchantments (true [use-experimental-damage] in config) / mcmmo are all premium, lastest.

this error Isn't related to AE?

**Comment by Insprill** [comment]
No

**Comment by Insprill** [comment]
@hkkongou any updates?

**Comment by hkkongou** [comment]
> @hkkongou any updates?

I have out the update version in one of my server within the network and so far , no invisible problem with mcmmo rupture or mythicmobs is reported from my players.

Which is good

And I am planning to re enable mcmmo rupture skill to all of my survival servers to see will this happen again , by increase the number of players to try on this issue

**Comment by Insprill** [comment]
awesome! glad to hear that. will keep this open for a little while longer to ensure that the setting fully fixes it.

**Comment by hkkongou** [comment]
bad news, players report having invisible mobs 
AE 8.6.10 experimental damage is on
https://www.youtube.com/watch?v=uaQywD-83Q4

**Comment by GC-spigot** [comment]
From what I can tell, fixed in next version
